### PR TITLE
video recorder as react component

### PR DESCRIPTION
**Pull request by YUzhva**
### JIRA:
**LR-200 Recorder custom UI**: https://ssu-jira.softserveinc.com/browse/LR-200

### Author.
@yuzhva

- [x] I have added link to this PR to the JIRA
- [x] I have logged time in JIRA.

### Reviewers.

> @anatoliiibrahimov
> @sergiy17
> @superNazo 

- [ ] @brouhaha88
- [x] @OKrain
- [x] @vovka

### Issue summary.

Custom UI with outer control elements like buttons, links etc. What we need: 
- Finish counter. 
- Start counter. 
- Record button. 
- Stop button. 
- Play button (?).
- Next question button. 

I would suggest to use ReactJS.

### Testing approach.

Manual testing of react componetns.
Jasmine test for AngularJS video_recorder page route.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 source 'https://rubygems.org'
-ruby "2.3.0"
+ruby '2.3.0'
```
Comment (vovka) [inline Gemfile]: Why do you all change these quotes?

File: app/assets/javascripts/react_app/components/Countdown.jsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    return (
+      <span style={secondsStyle}>
+        {seconds}
+      </span>
+    );
+  }
+});
+
+var renderCountdwown = function(seconds, autorun) {
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: What is this function for?

File: app/assets/javascripts/react_app/components/Stopwatch.jsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    return (
+      <span style={secondsStyle}>
+        {seconds}
+      </span>
+    );
+  }
+});
+
+var renderStopwatch = function(seconds, autorun) {
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Stopwatch.jsx]: What is this function for?

File: app/assets/javascripts/react_app/components/Countdown.jsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  },
+
+  start: function() {
+    this.getFlux().actions.startCountdown();
+  },
+
+  render: function() {
+    var countdown = Math.round(this.state.countdown);
+    var seconds = (countdown / 10).toFixed(1);
+    var secondsStyle = {color: 'red'};
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: Probably, you would better move styles in css.

File: app/assets/javascripts/react_app/components/Countdown.jsx (hunk 2 above)
Comment (YUzhva) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: renderCountdwown() - this is a function that we could call from our .html and render React class inside <div></div> tag

File: app/assets/javascripts/react_app/components/Countdown.jsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    if (this.props.autorun == true)
+      this.start(this.state.setTime);
+  },
+
+  start: function() {
+    this.getFlux().actions.startCountdown();
+  },
+
+  render: function() {
+    var countdown = Math.round(this.state.countdown);
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: Where this `countdown` state is being set?

File: app/assets/javascripts/react_app/components/Countdown.jsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    var seconds = (countdown / 10).toFixed(1);
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: Why do you divide it and cast type?

File: Gemfile (hunk 1 above)
Comment (YUzhva) [inline Gemfile]: Everywhere in our Gemfile we are using double quotes except this first line...

First line doesn't require convenience in next lines of Gemfile

File: app/assets/javascripts/react_app/components/Countdown.jsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
+var FluxMixin = Fluxxor.FluxMixin(React),
+  StoreWatchMixin = Fluxxor.StoreWatchMixin;
+
+var Countdown = React.createClass({
+  mixins: [FluxMixin, StoreWatchMixin("CountdownStore")],
+
+  getStateFromFlux: function() {
```
Comment (vovka) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: What is this method for?

File: app/assets/javascripts/react_app/components/Countdown.jsx (hunk 5 above)
Comment (YUzhva) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: in Flux, then we get all FLUX state with getStateFromFlux

File: app/assets/javascripts/react_app/components/Countdown.jsx (hunk 6 above)
Comment (YUzhva) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: Read in some example (=

File: app/assets/javascripts/react_app/components/Countdown.jsx (hunk 7 above)
Comment (YUzhva) [inline app/assets/javascripts/react_app/components/Countdown.jsx]: This is method from Fluxxor:
http://fluxxor.com/guides/quick-start.html

it get global states that we stored in out Sotores, so than we could use this.state from Store inside our react component

File: app/assets/javascripts/react_app/components/Stopwatch.jsx (hunk 3 above)
Comment (YUzhva) [inline app/assets/javascripts/react_app/components/Stopwatch.jsx]: This function helps us to render our react component with flux variable inside `<div></div>` tag at any .html page.

**Comment by OKrain** [comment]
:+1:

**Comment by vovka** [comment]
I allow mergin this without even tests:) in order to let @sergiy17 integrate recorder on his page. 
But my ask about moving business logic to stores is still valid.

File: spec/javascripts/angularjs_app/videoQuiz.routes_spec.js [hunk 8]
```diff
@@ -115,6 +115,42 @@ describe('videoQuizRoutes', function() {
     });
   });
 
+  describe('/video_recorder route', function() {
```
Comment (vovka) [inline spec/javascripts/angularjs_app/videoQuiz.routes_spec.js]: Codacy Issue found: ["describe" is not defined.](https://www.codacy.com/app/vovka/video_quiz/file/2613084302/issues/source?bid=3205144&fileBranchId=3305597#l118)

File: spec/javascripts/angularjs_app/videoQuiz.routes_spec.js [hunk 9]
```diff
@@ -115,6 +115,42 @@ describe('videoQuizRoutes', function() {
...
+    it('should have right template', function() {
+      var quizzesRoute = route.routes['/video_recorder'];
+      expect(quizzesRoute.templateUrl)
+        .toEqual('angularjs_app/shared/video_recorder/testingPage.html');
+    });
+  });
+});
+
+describe('videoQuizRoutes', function(){
+  var $route, $rootScope, $location, $httpBackend;
```
Comment (vovka) [inline spec/javascripts/angularjs_app/videoQuiz.routes_spec.js]: Codacy Issue found: ["$route" is defined but never used](https://www.codacy.com/app/vovka/video_quiz/file/2613084302/issues/source?bid=3205144&fileBranchId=3305597#l133)

File: spec/javascripts/angularjs_app/videoQuiz.routes_spec.js [hunk 10]
```diff
@@ -115,6 +115,42 @@ describe('videoQuizRoutes', function() {
...
+
+  beforeEach(function(){
+    module('videoQuiz');
+
+    inject(function($injector){
```
Comment (vovka) [inline spec/javascripts/angularjs_app/videoQuiz.routes_spec.js]: Codacy Issue found: ["inject" is not defined.](https://www.codacy.com/app/vovka/video_quiz/file/2613084302/issues/source?bid=3205144&fileBranchId=3305597#l138)

File: spec/javascripts/angularjs_app/videoQuiz.routes_spec.js [hunk 11]
```diff
@@ -115,21 +133,17 @@ describe('videoQuizRoutes', function() {
     });
   });
 
-  describe('/quizzes/:quizId show route', function() {
```
Comment (vovka) [inline spec/javascripts/angularjs_app/videoQuiz.routes_spec.js]: Codacy Issue found: ["describe" is not defined.](https://www.codacy.com/app/vovka/video_quiz/file/2613118862/issues/source?bid=3205144&fileBranchId=3305597#l136)

**Pull request merged**